# bacteriophage
Shell, R and Python scripts to reproduce the bioinformatics pre-precessing and statistical analyses described in our paper ([https://doi.org/10.1002/sim.10058](https://doi.org/10.1002/sim.10058)) "Multiblock Partial Least Squares and Rank Aggregation: Applications to Detection of Bacteriophages Associated with Antimicrobial Resistance in the Presence of Potential Confounding Factors" by Shoumi Sarkar, Samuel Anyaso-Samuel, Peihua Qiu and Somnath Datta. The programs are organized sequentially in the following folders:

`/01_PreprocessMetagenomicData` outlines steps on handling and preprocessing metagenomic data downloaded from the CAMDA server.

`/02a_RefSeq and 02b_ResFinder` go over obtaining abundances from the RefSeq and ResFinder databases respectively for bacteriophages and resistomes.

`/03_DownstreamAnalysis` describes the statistical methods implemented in our paper.

`/04_Supplementary` contains additional codes for the bioinformatics preprocessing and downstream analyses implemented in the Supplementary Material.
